enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,637 @@
+- Ensure that upgrade from a release running on the in-tree provider only to the out-of-tree succeeds, and the downgrade is also supported with proper CI tasks.
+- Ensure that upgrade from a release running on the in-tree provider only to the out-of-tree does not disrupt in-tree storage functionality with proper CI tasks.
+- Each provider should have a working CI implementation, which will assist in testing the provisioning and upgrades with out-of-tree support enabled.
+
+### Graduation Criteria
+
+#### Tech preview
+
+- `CCCMO` handles cluster installation with out-of-tree `CCM` and defaulting to in-tree provider for storage for at least one cloud-provider.
+
+#### Tech Preview -> GA
+
+- `CCCMO` handles basic tasks, such as initial installation templates for bootstrap and provision all of supported in-tree providers after installation.
+
+## FAQ
+
+Q: Can we disaster-recover a cluster without CCM running?
+
+- A: Yes we can, assuming the recovered state of the cluster has the `CCM` established properly, or it is still using in-tree implementation.
+
+Q: What is non-functional in a cluster during bootstrapping until CCM is up?
+
+- A: If CCM is not available, new nodes in the cluster will be left unschedulable with `node.cloudprovider.kubernetes.io/uninitialized` taint. This taint has `NoSchedule` effect, which will prevent most workloads from scheduling on the Node until CCM complete Node initialization.
+
+Q: Who talks to CCM?
+
+- A: CCM exposes health, readiness, metrics and config endpoints similar to most core Kubernetes controllers.
+These endpoints are accessed by the process host's Kubelet to ensure the process is running smoothly and Prometheus to gather metrics.
+No other components talk to the CCM directly. The CCM coordinates with other components by updating resources using the KAS similar to other controllers.
+
+Q: CCM relation to other openshift components, such as SDN and storage? How a non-operational CCM will affect cluster health, which components will take a hit?
+
+- A: A couple of components rely on CCM at different time of the cluster lifetime:
+  - `kubelet` dependency requires removal of taints from newly created Nodes. Without CCM, new nodes will not become schedulable, existing nodes are unaffected.
+  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down.
+  This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator).
+  Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.
```

Comment:
CCM is not required to implement routes if the CNI component (e.g. Calico) does it.  Will this use case be supported?

Reply:
According to the cloud interface implementation you could enable/disable parts of it. If your CCM works in such configuration from your testing, then it should work correctly in Openshift as well. But this question should be first discussed with networking team, we can't possibly know all details now.